### PR TITLE
SpatioTemporalClassifier, SpatioTemporalRegressor and refactored QuakeSet

### PR DESCRIPTION
**Pull request by DarthReca**
This is the draft implementation of the SpatioTemporalClassifier and the refactored QuakeSet. The regressor is currently WIP.

I have taken inspiration from #3280 but with some modifications (I still need to fully align the code with this pull request).

This is linked to both #3042 and #2382.

### CODE REVIEW

File: tests/tasks/test_spationtemporal_classification.py
Comment (adamjstewart) [inline tests/tasks/test_spationtemporal_classification.py:1]: Typo in filename

File: tests/conf/quakeset_classification.yaml
Comment (adamjstewart) [inline tests/conf/quakeset_classification.yaml:1]: Maybe call this `quakeset_classification.yaml` since we will also add a regression version in the future?

File: torchgeo/datamodules/quakeset.py [hunk 1]
```diff
@@ -34,9 +34,14 @@ def __init__(
         """
         super().__init__(QuakeSet, batch_size, num_workers, **kwargs)
         self.train_aug = K.AugmentationSequential(
-            K.Normalize(mean=self.mean, std=self.std),
```
Comment (adamjstewart) [inline torchgeo/datamodules/quakeset.py:37]: We still want normalization during training time, that should stay

File: torchgeo/trainers/spatiotemporal_classification.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+from ..datamodules import BaseDataModule
+from ..datasets import RGBBandsMissingError, unbind_samples
+from ..datasets.utils import Sample
+from ..models import ConvLSTM
+from .base import BaseTask
+from .mixins import ClassificationMixin
+
+
+class _ConvLSTMClassifier(nn.Module):
```
Comment (adamjstewart) [inline torchgeo/trainers/spatiotemporal_classification.py]: This is now removed from the other PR, let's wait until that gets merged and then update

Comment (DarthReca) [reply torchgeo/trainers/spatiotemporal_classification.py]: I have also made a change (uncomitted for now) using the TIMM [ClassifierHead](https://github.com/huggingface/pytorch-image-models/blob/main/timm/layers/classifier.py#L77); probably it's more stable than reinventing the wheel

**Comment by adamjstewart** [comment]
FYI, https://github.com/torchgeo/torchgeo/pull/3671 is finally merged! Can copy-n-paste this as the foundation for this PR.

File: torchgeo/trainers/spatiotemporal_classification.py (hunk 2 above)
Comment (DarthReca) [reply torchgeo/trainers/spatiotemporal_classification.py]: I have made a small change to the ConvLSTM head. If you believe it is fine or if you suggest any changes, I will update the tests, etc., once stable.

Comment (adamjstewart) [reply torchgeo/trainers/spatiotemporal_classification.py]: Does this work for regression too?

Comment (DarthReca) [reply torchgeo/trainers/spatiotemporal_classification.py]: Yep, it is simply the standard head builder of timm that can be customized a bit (for example, using a conv or linear). There are also some other versions based on MLP if I see well in the timm file.

**Comment by DarthReca** [comment]
Mostly done, I need to investigate what is wrong with ReadTheDocs

File: torchgeo/trainers/spatiotemporal_regression.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+import torch.nn as nn
+from torch import Tensor
+from torchmetrics import MeanAbsoluteError, MeanSquaredError, MetricCollection
+
+from ..datasets.utils import Sample
+from ..models import ConvLSTM
+from .base import BaseTask
+
+
+class SpatioTemporalRegressionTask(BaseTask):
```
Comment (robmarkcole) [inline torchgeo/trainers/spatiotemporal_regression.py]: Just to confirm, this differs from SpatioTemporalPixelwiseRegressionTask in https://github.com/torchgeo/torchgeo/pull/3086 right? I wonder if we can make this clearer, i.e. this task is for regressing a single patch level value

File: torchgeo/tasks/spatiotemporal_regression.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        num_filters: int = 3,
+        loss: Literal['mse', 'mae'] = 'mse',
+        lr: float = 1e-3,
+        patience: int = 10,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize a new SpatioTemporalRegressionTask instance.
+
+        Args:
+            model: Video model name. Only ``'convlstm'`` is currently supported.
```
Comment (robmarkcole) [inline torchgeo/tasks/spatiotemporal_regression.py]: @adamjstewart didn't we decide agains this `video` naming approach?

File: torchgeo/trainers/spatiotemporal_regression.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+        * :class:`~torchmetrics.MeanSquaredError`: The average of the squared
+        differences between the predicted and actual values (MSE) and its
+        square root (RMSE). Lower values are better.
+        * :class:`~torchmetrics.MeanAbsoluteError`: The average of the absolute
+        differences between the predicted and actual values (MAE).
+        Lower values are better.
+        """
+        metrics = MetricCollection(
+            # https://github.com/astral-sh/ty/issues/2985
```
Comment (robmarkcole) [inline torchgeo/trainers/spatiotemporal_regression.py]: This issue is closed so why is it referenced?

File: torchgeo/models/convlstm.py [hunk 6]
```diff
@@ -134,6 +136,8 @@ def __init__(
             num_classes: Optional number of segmentation classes for an attached
                 prediction head.
             head_kernel_size: Kernel size for the optional segmentation head.
+            convolutional_head: If ``True``, uses a convolutional head for
+                segmentation instead of a fully connected head.
```
Comment (robmarkcole) [inline torchgeo/models/convlstm.py]: `a fully connected head` is the appropriate choice for image-level prediction? could be a useful comment to include

File: torchgeo/tasks/spatiotemporal_classification.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
+# Copyright (c) TorchGeo Contributors. All rights reserved.
+# Licensed under the MIT License.
+
+"""Trainers for spatiotemporal regression."""
```
Comment (robmarkcole) [inline torchgeo/tasks/spatiotemporal_classification.py]: I think it would be useful to clarify 'at image level' to avoid confusion with pixel level tasks

**Comment by robmarkcole** [comment]
This LGTM - would be nice to see a full training result with this new trainer on the dataset?

File: torchgeo/tasks/spatiotemporal_regression.py (hunk 4 above)
Comment (adamjstewart) [reply torchgeo/tasks/spatiotemporal_regression.py]: Yes, we typically refer to "video" input as "spatiotemporal" input. As long as it isn't a user-facing class/parameter name, it doesn't matter that much though.

File: torchgeo/trainers/spatiotemporal_regression.py (hunk 5 above)
Comment (adamjstewart) [reply torchgeo/trainers/spatiotemporal_regression.py]: Probably copy-n-paste. Shouldn't need to add any metrics if you inherit from RegressionMixin.

File: torchgeo/trainers/spatiotemporal_regression.py (hunk 3 above)
Comment (adamjstewart) [reply torchgeo/trainers/spatiotemporal_regression.py]: #3749 will probably help make this more clear.

Comment (DarthReca) [reply torchgeo/trainers/spatiotemporal_regression.py]: Yes, correct. They are both creating a single real or categorical value per timeseries

File: torchgeo/tasks/spatiotemporal_regression.py (hunk 4 above)
Comment (DarthReca) [reply torchgeo/tasks/spatiotemporal_regression.py]: Let me copy the new naming from the accepted MR (probably this was taken from the old version)

File: torchgeo/trainers/spatiotemporal_regression.py (hunk 5 above)
Comment (DarthReca) [reply torchgeo/trainers/spatiotemporal_regression.py]: Ok, thanks. I have missed the RegressionMixin

Comment (adamjstewart) [reply torchgeo/trainers/spatiotemporal_regression.py]: That should also fix the RtD test failure?

**Comment by adamjstewart** [comment]
Can you rebase? We moved all trainers to tasks.

**Comment by DarthReca** [comment]
Yup, after coming back from vacation I will finish this PR

**Comment by robmarkcole** [comment]
I think there could be some confusion between SpatioTemporalPixelwiseRegression added in https://github.com/torchgeo/torchgeo/pull/3086 and SpatioTemporalRegression. Perhaps SpatioTemporalPatchwiseRegression would be clearer @adamjstewart ?

**Comment by adamjstewart** [comment]
There hasn't been confusion between our existing Regression and PixelwiseRegression has there?

**Comment by robmarkcole** [comment]
@adamjstewart I guess not, as you would know if there has :-)

**Comment by adamjstewart** [comment]
I would know if someone complained to me. That's why we're planning a user survey, to get feedback on stuff like this.

File: torchgeo/trainers/spatiotemporal_regression.py (hunk 5 above)
Comment (DarthReca) [reply torchgeo/trainers/spatiotemporal_regression.py]: seems solved

**Comment by DarthReca** [comment]
@adamjstewart @robmarkcole Meanwhile, I am testing the training, I am not sure if the update to convlstm sounds to you and the uncovered line can be removed or exists for a reason.
